framework_agreement/model/portfolio.py

```diff
@@ -35,15 +42,212 @@ def get_from_supplier(self, supplier):
+                'name': '{}: Main version'.format(self.name),
+                'date_start': self.start_date,
+                'date_end': self.end_date,
+                'items_id': [(0, 0, {
+                    'name': '{}: Main rule'.format(self.name),
+                    'base': -2,
+                    'use_agreement_prices': True,
+                })],
+            })],
+        })
+
+        for product_line in self.line_ids:
+            self.env['product.supplierinfo'].create({
+                'name': self.supplier_id.id,
+                'agreement_pricelist_id': new_agreement.id,
+                'product_tmpl_id': product_line.product_id.product_tmpl_id.id,
+                'product_name': product_line.product_id.name,
+                'product_code': product_line.product_id.code,
+                'pricelist_ids': [(0, 0, {
+                    'min_quantity': 1.0,
+                    'price': 0.0,
+                })],
+            })
+
+    @api.multi
+    def get_line_for_product(self, product, quantity=0):
+        self.ensure_one()
+        for product_line in self.line_ids:
+            if (product_line.product_id == product and
+                    product_line.quantity >= quantity):
+                return product_line
+        return False
+
+    @api.multi
+    def is_suitable_for(self, date, product, quantity=0):
+        # we prefer doing a rough check and let the user make their choices
+        # instead of risking to limit things too much for a small discrepancy.
+        # Think of an agreement that would be good a few days later, or with a
+        # slightly different incoterm.
+        return self.state in ('running', 'future')
+
+    @api.multi
+    def is_valid_at_date(self, proposed_date):
+        self.ensure_one()
+        start = datetime.strptime(self.start_date, DATE_FORMAT)
+        is_valid = start < proposed_date
+
+        if self.end_date:
+            end = datetime.strptime(self.end_date, DATE_FORMAT)
+            is_valid = is_valid and proposed_date < end
+
+        return is_valid
+
+
+class AgreementProductLine(models.Model):
+    _name = 'agreement.product.line'
+
+    portfolio_id = fields.Many2one('framework.agreement.portfolio',
+                                   required=True)
+    product_id = fields.Many2one('product.product', 'Product', required=True)
+    quantity = fields.Integer('Negociated quantity', required=True)
+    available_quantity = fields.Integer(
+        compute='_compute_available_qty',
+        string='Available quantity',
+        store=True,
+        # this is stored because we do searches on it. However, the method
+        # might be called more often than necessary (specifically, when
+        # computing the state of the portfolio).
+    )
+
+    @api.depends(
+        'quantity',
+        'product_id',
+        'portfolio_id',
+        'portfolio_id.purchase_ids',
+        'portfolio_id.purchase_ids.portfolio_id',
+        'portfolio_id.purchase_ids.state',
```

Comment:
`portfolio_id.purchase_ids.portfolio_id` aleady implies
`portfolio_id.purchase_ids` and `portfolio_id`